File: shared/src/main/java/xyz/klinker/messenger/shared/service/jobs/CleanupOldMessagesJob.java
package xyz.klinker.messenger.shared.service.jobs;

import android.app.AlarmManager;
import android.app.IntentService;
import android.app.PendingIntent;
import android.app.job.JobInfo;
import android.app.job.JobParameters;
import android.app.job.JobScheduler;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;

import java.util.Date;

import xyz.klinker.messenger.shared.data.DataSource;
import xyz.klinker.messenger.shared.data.Settings;
import xyz.klinker.messenger.shared.util.TimeUtils;

public class CleanupOldMessagesJob extends BackgroundJob {

    private static final int JOB_ID = 1330;

    @Override
    protected void onRunJob(JobParameters parameters) {
        long timeout = Settings.INSTANCE.getCleanupMessagesTimeout();
        if (timeout > 0) {
            DataSource.INSTANCE.cleanupOldMessages(this, System.currentTimeMillis() - timeout);
        }

        scheduleNextRun(this);
    }

    public static void scheduleNextRun(Context context) {
        ComponentName component = new ComponentName(context, CleanupOldMessagesJob.class);
        JobInfo.Builder builder = new JobInfo.Builder(JOB_ID, component)
                .setMinimumLatency(TimeUtils.INSTANCE.millisUntilHourInTheNextDay(3)) // 3 AM
                .setRequiresCharging(true)
                .setRequiresDeviceIdle(true);

        JobScheduler jobScheduler = (JobScheduler) context.getSystemService(Context.JOB_SCHEDULER_SERVICE);
        jobScheduler.schedule(builder.build());
    }
}

File: shared/src/main/java/xyz/klinker/messenger/shared/service/jobs/CleanupOldMessagesJob.kt
package xyz.klinker.messenger.shared.service.jobs

import android.content.Context
import com.firebase.jobdispatcher.*
import xyz.klinker.messenger.shared.data.DataSource
import xyz.klinker.messenger.shared.data.Settings
import xyz.klinker.messenger.shared.util.TimeUtils

class CleanupOldMessagesJob : SimpleJobService() {
    override fun onRunJob(job: JobParameters): Int {
        val timeout = Settings.cleanupMessagesTimeout
        if (timeout > 0) {
            DataSource.cleanupOldMessages(this, TimeUtils.now - timeout)
        }

        scheduleNextRun(this)
        return JobService.RESULT_SUCCESS
    }

    companion object {
        private val JOB_ID = "cleanup-old-messages"

        fun scheduleNextRun(context: Context) {
            val dispatcher = FirebaseJobDispatcher(GooglePlayDriver(context))
            val time = (TimeUtils.millisUntilHourInTheNextDay(3).toLong() / 1000).toInt()
            val myJob = dispatcher.newJobBuilder()
                    .setService(CleanupOldMessagesJob::class.java)
                    .setTag(JOB_ID)
                    .setRecurring(false)
                    .setLifetime(Lifetime.FOREVER)
                    .setTrigger(Trigger.executionWindow(time, time + (TimeUtils.MINUTE.toInt() / 1000 * 5)))
                    .setReplaceCurrent(true)
                    .build()

            dispatcher.mustSchedule(myJob)
        }
    }
}
